Refactor attributes and improve argument parsing

mudge opened the pull request:
- **Remove redundant stripping of whitespace**
- **Expand usage of argparse to handle CLI options**
- **Refactor attribute handling**

mudge committed:
Remove redundant stripping of whitespace

Comparing the resulting AsciiDoc with and without this extra stripping
of trailing whitespace on the current Raspberry Pi Pico C/C++ SDK
databook results in no change in output.

mudge committed:
Expand usage of argparse to handle CLI options

Rather than using optional string arguments for both input and output files,
use more of Python's argparse library to handle a single, mandatory,
positional file argument (which now returns a handle to read the file).

This is a breaking change as the "-f" argument is now a positional one
so any downstream usage will need to be updated.

Note we don't use `argparse.FileType("w")` for the output file as that
would cause the given file to be created or truncated as soon as the
program runs when we only want it to produce output if successful.

mudge committed:
Refactor attribute handling

Rather than passing around an AsciiDoc string of attributes, pass around
a dictionary that is only converted to AsciiDoc at the very end when
passed to the `title` helper.

Note I considered switching to using the role shorthand for multiple
roles (e.g. `.h6.contextspecific` rather than `role=h6 contextspecific`)
but there is downstream code that relies on matching the existing
format. Similarly, the order of the `tag` and `type` attributes is
currently relied upon by client code so these have to be special cased
for now.

This also removes the dedicated Title Node class in favour of handling
title rendering in the parent element (in this case, the Sect Node).

mudge committed:
Remove redundant second argument to get

lurch commented on cli.py:
```diff
@@ -8,40 +7,43 @@
-        "-o", "--output", help="The path of the output file", default=None
+        "-o",
+        "--output",
+        type=argparse.FileType("w", encoding="utf-8"),
```
Does this usage of `argparse.FileType` (never seen that before - funky!) mean that a zero-length file will get created as soon as the program runs, regardless of whether the later XML-parsing throws any errors?
Trying to avoid that situation is exactly why I created #4 in the first place :wink:

mudge replied:
You're right: it will eagerly open the file for writing, blowing away any existing contents even if the AsciiDoc conversion fails. Let me revise this.

mudge replied:
Fixed in https://github.com/raspberrypi/doxygentoasciidoc/pull/8/commits/56ee954468864fb781d6f9fa3b686fdcce6bfdad

lurch commented on helpers.py:
```diff
@@ -19,19 +19,40 @@ def sanitize(identifier):
+    attrlist = []
+
+    if "id" in attributes:
+        attrlist.append(f"#{sanitize(attributes.pop('id'))}")
```
Haha, I was _about_ to suggest that it might be worth making `attrlist` a dict instead of a list, but then I spotted that this first item **isn't** of the form "xxx=yyy".

mudge replied:
If we can break the dependency on matching the `role=h6 contextspecific` substring, we can adopt a more concise syntax here and use the [AsciiDoc shorthand](https://docs.asciidoctor.org/asciidoc/latest/attributes/positional-and-named-attributes/#block-style-and-attribute-shorthand):

```asciidoc
[#group_foo.h6.contextspecific,tag=PICO_2040,type=PICO_2040]
```

lurch commented on helpers.py:
```diff
@@ -19,19 +19,40 @@ def sanitize(identifier):
+    if attributes is None:
+        attributes = {}
+
+    output = []
```
Only a small nit, but is it worth moving this declaration of `output` down to line 49, to make it more obvious that it doesn't get modified until near the end of the function?

mudge replied:
Fixed in https://github.com/raspberrypi/doxygentoasciidoc/pull/8/commits/d354949fea193faa41db9748f465178068e2fd47

lurch commented on nodes.py:
```diff
@@ -762,7 +746,7 @@ def to_asciidoc(self, **kwargs):
 class ComputeroutputNode(Node):
     def to_asciidoc(self, **kwargs):
-        return "".join(("`", super().to_asciidoc(**kwargs).rstrip(), "`"))
+        return "".join(("`", super().to_asciidoc(**kwargs), "`"))
```
Would it be cleaner to use an f-string here? Something like:
```python3
return f"`{super().to_asciidoc(**kwargs)}`"
```

mudge replied:
Fixed in https://github.com/raspberrypi/doxygentoasciidoc/pull/8/commits/8412caa33f97d1fee13cb4db3de135f3658b78e5

lurch commented:
:+1:  to keeping things as dicts/lists for as long as possible, rather than prematurely converting to strings.

mudge committed:
Prefer f-strings where possible

Rather than joining tuples on the empty string, use f-strings to build
up simple strings with interpolated expressions.

lurch commented on nodes.py:
```diff
@@ -648,52 +629,55 @@ def to_asciidoc(self, **kwargs):
         previous_node = self.previous_node()
         next_node = self.next_node()
-        if self.node.get("kind", "") == "see":
+        if self.node.get("kind") == "see":
```
Looks like lines 665 and 668 could be converted from `if` to `elif` ? And given that they're _all_ doing `self.node.get("kind")` perhaps its worth doing `kind = self.node.get("kind")` above line 648?

mudge replied:
As there is an early return on line 666 😈, pylint will complain if you switch to an `elif`:

> Unnecessary "elif" after "return", remove the leading "el" from "elif"

mudge replied:
Memoized `self.node.get("kind")` in https://github.com/raspberrypi/doxygentoasciidoc/pull/8/commits/06370ed29ce495af4fb001f61aff36efaa367b66

lurch reviewed (commented):
One small nitpick but otherwise LGTM!

mudge committed:
Memoize kind in Simplesect Node

Rather than repeatedly loading the kind from the node's attributes,
fetch it once and then switch on the resulting value multiple times.
